lib/SlackBot.js

```diff
@@ -225,113 +225,253 @@ function Slackbot(configuration) {
+            if (slack_botkit.config.require_delivery) {
+                // loop through all active conversations this bot is having
+                // and mark messages in conversations as delivered = true
+                for (var t = 0; t < slack_botkit.tasks.length; t++) {
+                    if (slack_botkit.tasks[t].isActive()) {
+                        for (var c = 0; c < slack_botkit.tasks[t].convos.length; c++) {
+                            for (var s = 0; s < slack_botkit.tasks[t].convos[c].sent.length; s++) {
+                                if (slack_botkit.tasks[t].convos[c].sent[s].api_response && slack_botkit.tasks[t].convos[c].sent[s].api_response.id == message.reply_to) {
+                                    slack_botkit.tasks[t].convos[c].sent[s].delivered = true;
+                                    slack_botkit.tasks[t].convos[c].sent[s].api_response.ts = message.ts;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            return false;
+        }
+
+        next();
+    });
+
+
+
+    slack_botkit.middleware.categorize.use(function(bot, message, next) {
+
+        var mentionSyntax = '<@' + bot.identity.id + '(\\|' + bot.identity.name.replace('.', '\\.') + ')?>';
+        var mention = new RegExp(mentionSyntax, 'i');
+        var direct_mention = new RegExp('^' + mentionSyntax, 'i');
+
+        if ('message' == message.type) {
+
+            if (message.text) {
+                message.text = message.text.trim();
+            }
+
+            // set up a couple of special cases based on subtype
+            if (message.subtype && message.subtype == 'channel_join') {
+                // someone joined. maybe do something?
+                if (message.user == bot.identity.id) {
+                    message.type = 'bot_channel_join';
+                } else {
+                    message.type = 'user_channel_join';
+                }
+            } else if (message.subtype && message.subtype == 'group_join') {
+                // someone joined. maybe do something?
+                if (message.user == bot.identity.id) {
+                    message.type = 'bot_group_join';
+                } else {
+                    message.type = 'user_group_join';
+                }
+            } else if (message.subtype) {
+                message.type = message.subtype;
+            } else if (message.channel.match(/^D/)) {
+                // this is a direct message
+                if (message.user == bot.identity.id && message.bot_id) {
+                    return false;
+                }
+                if (!message.text) {
+                    // message without text is probably an edit
+                    return false;
+                }
+
+                // remove direct mention so the handler doesn't have to deal with it
+                message.text = message.text.replace(direct_mention, '')
```

Comment:
text.trim() might be nicer than the regex to replace whitespace since it reads a little more clearly